### ISSUE TITLE
Throwing from Base.show(::IO, ::Type{...}) kills the REPL

### ISSUE THREAD

**Issue by tkf** [JuliaLang/julia#29428]
I checked it with 1.0.0 and 1.1.0-DEV.334.

```julia
julia> struct Spam end

julia> Base.show(::IO, ::Type{Spam}) = error("nope")

julia> Spam
Error showing value of type DataType:
ERROR: nope
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] show(::IOContext{REPL.Terminals.TTYTerminal}, ::Type{SYSTEM: show(lasterr) caused an error
ErrorException("nope")

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] show(::IOContext{REPL.Terminals.TTYTerminal}, ::Type{ERROR: nope
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] show(::IOContext{REPL.Terminals.TTYTerminal}, ::Type{fatal: error thrown and no exception handler available.
ErrorException("nope")
rec_backtrace at /home/takafumi/repos/watch/julia/src/stackwalk.c:94
record_backtrace at /home/takafumi/repos/watch/julia/src/task.c:249
jl_throw at /home/takafumi/repos/watch/julia/src/task.c:580
error at ./error.jl:33
show at ./REPL[2]:1
jl_fptr_trampoline at /home/takafumi/repos/watch/julia/src/gf.c:1843
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
show_datatype at ./show.jl:526
show at ./show.jl:436
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
print at ./strings/io.jl:31
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
print at ./strings/io.jl:42
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
show_tuple_as_call at ./show.jl:1490
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
show_spec_linfo at ./stacktraces.jl:261
#show#9 at ./stacktraces.jl:272
unknown function (ip: 0x7f99e80ea7c3)
#show at ./none:0 [inlined]
#show_trace_entry#630 at ./errorshow.jl:471
unknown function (ip: 0x7f99e80ea0ec)
#show_trace_entry at ./none:0
unknown function (ip: 0x7f99e80e9e28)
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
show_backtrace at ./errorshow.jl:574
#showerror#612 at ./errorshow.jl:79
unknown function (ip: 0x7f99e80e7357)
jl_fptr_trampoline at /home/takafumi/repos/watch/julia/src/gf.c:1843
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
showerror at ./errorshow.jl:74 [inlined]
display_error at ./client.jl:99
jl_fptr_trampoline at /home/takafumi/repos/watch/julia/src/gf.c:1843
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
display_error at ./client.jl:102
jl_fptr_trampoline at /home/takafumi/repos/watch/julia/src/gf.c:1843
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
jl_apply at /home/takafumi/repos/watch/julia/src/julia.h:1559 [inlined]
jl_f__apply at /home/takafumi/repos/watch/julia/src/builtins.c:556
jl_f__apply_latest at /home/takafumi/repos/watch/julia/src/builtins.c:594
#invokelatest#1 at ./essentials.jl:697 [inlined]
invokelatest at ./essentials.jl:696 [inlined]
_start at ./client.jl:423
jl_apply_generic at /home/takafumi/repos/watch/julia/src/gf.c:2198
jl_apply at /home/takafumi/repos/watch/julia/ui/../src/julia.h:1559 [inlined]
true_main at /home/takafumi/repos/watch/julia/ui/repl.c:112
main at /home/takafumi/repos/watch/julia/ui/repl.c:233
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
_start at ./julia (unknown line)
```

**Comment by KristofferC** [comment]
https://github.com/JuliaLang/julia/issues/22363 etc

**Comment by tkf** [comment]
@KristofferC Thanks for the reference. I searched the issue but I couldn't find it.  But I'm not sure if it's a dup of #22363 as the error is different.  It's a segfault in #22363 and a lack of exception handler here.  This issue is more like a dup of #21119.

I think this can be fixed purely in Julia (somewhere in stacktrace printing?).  I suggested [a fix](https://github.com/JuliaPy/PyCall.jl/blob/6b297e1018fa36a007bdca0d8c4ce82b9d43d26e/src/exception.jl#L126-L164) in PyCall.jl to avoid a similar issue here: https://github.com/JuliaPy/PyCall.jl/pull/583.


> I never condone defining show for particular types, which goes beyond "type piracy" and into "type treason" --- https://github.com/JuliaLang/julia/issues/22363#issuecomment-308472359

@JeffBezanson I thought it was a part of the interface (like `convert(::Type{MyType}, ...)`) since Enums does it:

https://github.com/JuliaLang/julia/blob/ebfbe62ef1d624b1f7103c7f67ee635eadad2f21/base/Enums.jl#L161-L172

If it is a part of the interface, doesn't it make sense to avoid user's mistake to bring down the whole Julia process?  (Or maybe by "particular types" you mean that "types you do not own"?)

I'm bringing this up here since @KristofferC closed #21119 since

> Overloading show for types is knows to be iffy --- https://github.com/JuliaLang/julia/issues/21119#issuecomment-355847371

**Comment by tkf** [comment]
Making `show(::IO, ::StackFrame)` slightly more error-tolerant (see blow) seems to solve this problem. Rendering `Spam` still throws (which is the correct behavior) but it does not kill `julia`.

```julia
julia> using Base.StackTraces: StackFrame, empty_sym, show_spec_linfo

julia> function Base.show(io::IO, frame::StackFrame; full_path::Bool=false)
           try
               show_spec_linfo(io, frame)
           catch
               print(io, "[ERROR RENDERING INFO]")  # added
           end
           if frame.file !== empty_sym
               file_info = full_path ? string(frame.file) : basename(string(frame.file))
               print(io, " at ")
               Base.with_output_color(get(io, :color, false) && get(io, :backtrace, false) ? Base.stackframe_lineinfo_color() : :nothing, io) do io
                   print(io, file_info, ":")
                   if frame.line >= 0
                       print(io, frame.line)
                   else
                       print(io, "?")
                   end
               end
           end
           if frame.inlined
               print(io, " [inlined]")
           end
       end

julia> struct Spam end

julia> Base.show(::IO, ::Type{Spam}) = error("nope")

julia> Spam
Error showing value of type DataType:
ERROR: nope
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] show(::IOContext{REPL.Terminals.TTYTerminal}, ::Type{[ERROR RENDERING INFO] at ./REPL[4]:1
 [3] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Type) at ./sysimg.jl:194
 [4] display(::REPL.REPLDisplay, ::MIME{Symbol("text/plain")}, ::Any) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:131
 [5] display(::REPL.REPLDisplay, ::Any) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:135
 [6] display(::Any) at ./multimedia.jl:287
 [7] #invokelatest#1 at ./essentials.jl:697 [inlined]
 [8] invokelatest at ./essentials.jl:696 [inlined]
 [9] print_response(::IO, ::Any, ::Any, ::Bool, ::Bool, ::Any) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:154
 [10] print_response(::REPL.AbstractREPL, ::Any, ::Any, ::Bool, ::Bool) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:139
 [11] (::getfield(REPL, Symbol("#do_respond#40")){Bool,getfield(REPL, Symbol("##50#59")){REPL.LineEditREPL,REPL.REPLHistoryProvider},REPL.LineEditREPL,REPL.LineEdit.Prompt})(::Any, ::Any, ::Any) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:708
 [12] #invokelatest#1 at ./essentials.jl:697 [inlined]
 [13] invokelatest at ./essentials.jl:696 [inlined]
 [14] run_interface(::REPL.Terminals.TextTerminal, ::REPL.LineEdit.ModalInterface, ::REPL.LineEdit.MIState) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/LineEdit.jl:2261
 [15] run_frontend(::REPL.LineEditREPL, ::REPL.REPLBackendRef) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:1029
 [16] run_repl(::REPL.AbstractREPL, ::Any) at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:191
 [17] (::getfield(Base, Symbol("##719#721")){Bool,Bool,Bool,Bool})(::Module) at ./logging.jl:311
 [18] #invokelatest#1 at ./essentials.jl:697 [inlined]
 [19] invokelatest at ./essentials.jl:696 [inlined]
 [20] macro expansion at ./logging.jl:308 [inlined]
 [21] run_main_repl(::Bool, ::Bool, ::Bool, ::Bool, ::Bool) at ./client.jl:330
 [22] exec_options(::Base.JLOptions) at ./client.jl:242
 [23] _start() at ./client.jl:421
```

**Comment by JeffBezanson** [comment]
The first `show` method for Enums shouldn't be there; I don't know why it's there since it does the exact same thing as the usual method. Defining the 3-argument form of `show` is a bit better, since it's much less likely to be used by lower-level code.

I guess it's ok to apply a patch like yours to make stacktrace printing more robust. But really, if you add such a low level method and it throws an error, shouldn't you just fix it?

Also a duplicate of #28983.

**Comment by tkf** [comment]
I don't need to define "low-level" show method for types *myself* to invoke this problem:

```julia
struct Unprintable end
Base.show(::IO, ::Unprintable) = error("show(::IO, ::Unprintable) called")
f(::Val{Unprintable()}) = error("f(::Val{Unprintable()}) called")
f(Val(Unprintable()))
```

(Note: It seems copy-and-pasting `"f(Val(Unprintable()))\n"` into REPL doesn't work. You have to insert `"f(Val(Unprintable()))"` and then hit enter.)

The problem is that it's so easy to touch low-level mechanism in Julia (which is, of course, the flip side of great features that Julia provides).  I think it helps users if Julia's stacktrace printer is a bit more paranoid.  But I think my quick hack above is bad since it hides the bug and users loose a chance to fix it.  I think it must report nested exception:

```julia
function Base.show(io::IO, frame::StackFrame; full_path::Bool=false)
    try
        show_spec_linfo(io, frame)
    catch ex
        println(io, "[ERROR RENDERING INFO]")
        @error "`show_spec_linfo` failed" exception=(ex, catch_backtrace())
    end
    ...
end
```

This implementation is of course not great since it could cause a stack overflow.  (Also, `@error` is not `io`-aware.)  To do it right I think there has to be a safer stacktrace printer and use it in the nested stacktrace (maybe using `IOContext`?).

At this point, I'm not super sure if this is the right way to go.  It may be over-paranoid to prepare for such a corner case.  But if it was a minor problem then there won't be multiple issues reporting it?

**Comment by tkf** [comment]
This seems to be fixed since 1.1.0 (thanks to @c42f, I guess?)

**Comment by c42f** [comment]
It's possible that I fixed this as a side effect of changing the way errors are displayed in the REPL, though I didn't fix it knowingly :-)
